curvefs/src/metaserver/heartbeat.cpp

```diff
@@ -342,13 +420,150 @@ void Heartbeat::HeartbeatWorker() {
             continue;
         }
 
-        // TODO(cw123): add schedule and execute heartbeat info
-
+        taskExecutor_->ExecTasks(resp);
         waitInterval_.WaitForNextExcution();
     }
 
     LOG(INFO) << "Heartbeat worker thread stopped.";
 }
 
+HeartbeatTaskExecutor::HeartbeatTaskExecutor(CopysetNodeManager* mgr,
+                                             const butil::EndPoint& endpoint)
+    : copysetMgr_(mgr), ep_(endpoint) {}
+
+void HeartbeatTaskExecutor::ExecTasks(const HeartbeatResponse& response) {
+    for (auto& conf : response.needupdatecopysets()) {
+        ExecOneTask(conf);
+    }
+}
+
+void HeartbeatTaskExecutor::ExecOneTask(const CopySetConf& conf) {
+    auto copyset = copysetMgr_->GetCopysetNode(conf.poolid(), conf.copysetid());
+    if (!copyset) {
+        LOG(WARNING) << "Failed to find copyset: " << CopysetName(conf);
+        return;
+    }
+
+    if (NeedPurge(conf)) {
+        DoPurgeCopyset(conf.poolid(), conf.copysetid());
+        return;
+    }
+
+    const auto epochInCopyset = copyset->GetConfEpoch();
+    if (conf.epoch() != epochInCopyset) {
+        LOG(WARNING) << "Config change epoch: " << conf.epoch()
+                     << " isn't same as current: " << epochInCopyset
+                     << ", copyset: " << copyset->Name()
+                     << ", refuse config change";
+        return;
+    }
+
+    switch (conf.type()) {
+        case ConfigChangeType::TRANSFER_LEADER:
+            DoTransferLeader(copyset, conf);
+            break;
+        case ConfigChangeType::ADD_PEER:
+            DoAddPeer(copyset, conf);
+            break;
+        case ConfigChangeType::REMOVE_PEER:
+            DoRemovePeer(copyset, conf);
+            break;
+        case ConfigChangeType::CHANGE_PEER:
+            DoChangePeer(copyset, conf);
+            break;
+        default:
+            LOG(ERROR) << "unexpected, type: " << conf.type();
+            break;
+    }
+}
+
+void HeartbeatTaskExecutor::DoTransferLeader(CopysetNode* node,
+                                             const CopySetConf& conf) {
+    LOG(INFO) << "Transferring leader to " << conf.configchangeitem().address()
+              << " of copyset: " << node->Name();
+
+    if (!IsAllCopysetsLoadFinished(conf.configchangeitem())) {
```

Comment:
Why not send the change request directly? When the copyset is not loaded, just discard it.

Reply:
fix